flink/src/test/java/org/apache/iceberg/flink/source/BoundedTableFactory.java

```diff
@@ -0,0 +1,151 @@
+import org.apache.flink.types.RowKind;
+import org.apache.iceberg.flink.util.FlinkCompatibilityUtil;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableSet;
+
+public class BoundedTableFactory implements DynamicTableSourceFactory {
+  private static final AtomicInteger DATA_SET_ID = new AtomicInteger(0);
+  private static final Map<String, List<List<Row>>> DATA_SETS = new HashMap<>();
+
+  private static final ConfigOption<String> DATA_ID = ConfigOptions.key("data-id").stringType().noDefaultValue();
+
+  public static String registerDataSet(List<List<Row>> dataSet) {
+    String dataSetId = String.valueOf(DATA_SET_ID.incrementAndGet());
+    DATA_SETS.put(dataSetId, dataSet);
+    return dataSetId;
+  }
+
+  public static void clearDataSets() {
+    DATA_SETS.clear();
+  }
+
+  @Override
+  public DynamicTableSource createDynamicTableSource(Context context) {
+    TableSchema tableSchema = TableSchemaUtils.getPhysicalSchema(context.getCatalogTable().getSchema());
+
+    Configuration configuration = Configuration.fromMap(context.getCatalogTable().getOptions());
+    String dataId = configuration.getString(DATA_ID);
+    Preconditions.checkArgument(DATA_SETS.containsKey(dataId),
+        "data-id %s does not found in registered data set.", dataId);
+
+    return new BoundedTableSource(DATA_SETS.get(dataId), tableSchema);
+  }
+
+  @Override
+  public String factoryIdentifier() {
+    return "BoundedSource";
+  }
+
+  @Override
+  public Set<ConfigOption<?>> requiredOptions() {
+    return ImmutableSet.of();
+  }
+
+  @Override
+  public Set<ConfigOption<?>> optionalOptions() {
+    return ImmutableSet.of(DATA_ID);
+  }
+
+  private static class BoundedTableSource implements ScanTableSource {
+
+    private final List<List<Row>> elementsPerCheckpoint;
+    private final TableSchema tableSchema;
+
+    private BoundedTableSource(List<List<Row>> elementsPerCheckpoint, TableSchema tableSchema) {
+      this.elementsPerCheckpoint = elementsPerCheckpoint;
+      this.tableSchema = tableSchema;
+    }
+
+    private BoundedTableSource(BoundedTableSource toCopy) {
+      this.elementsPerCheckpoint = toCopy.elementsPerCheckpoint;
+      this.tableSchema = toCopy.tableSchema;
+    }
+
+    @Override
+    public ChangelogMode getChangelogMode() {
+      return ChangelogMode.newBuilder()
+          .addContainedKind(RowKind.INSERT)
+          .addContainedKind(RowKind.DELETE)
+          .addContainedKind(RowKind.UPDATE_BEFORE)
+          .addContainedKind(RowKind.UPDATE_AFTER)
+          .build();
+    }
+
+    @Override
+    public ScanRuntimeProvider getScanRuntimeProvider(ScanContext runtimeProviderContext) {
+      return new DataStreamScanProvider() {
+        @Override
+        public DataStream<RowData> produceDataStream(StreamExecutionEnvironment env) {
+          SourceFunction<Row> source = new BoundedTestSource<>(elementsPerCheckpoint);
```

Comment:
Let's say `checkpointToAwait = numCheckpointsComplete.get() + delta`,   in fact the value of `delta` should not affect the final table records because we only need to make sure that there will be exactly `elementsPerCheckpoint.size()`  checkpoints to emit each records buffer from the original `elementsPerCheckpoint`.   Even if the checkpoints that emitted results are not continuous, the correctness of the data should not be affected in the end.   Setting the `delta` to be 2 is introduce the variable that produce un-continuous checkpoints.

Reply:
so we want the pattern of checkpoint cycles like this? Only small downside is slightly longer runtime. Now we need to wait for 6 checkpoints instead of 3.
```
insert batch 1, idle, insert batch 2, idle, insert batch 3, idle, ...
```

Regardless, it is helpful to add some comments to explain the decision. It is very difficult to know the intention of the magic number of 2.